### PR TITLE
Add `limitRequest` operator

### PR DESCRIPTION
**Pull request by simonbasle**
See #879

### CODE REVIEW

File: reactor-core/src/main/java/reactor/core/publisher/FluxLimitRequest.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+		return super.scanUnsafe(key);
+	}
+
+	static class FluxLimitRequestSubscriber<T> implements InnerOperator<T, T> {
+
+		final CoreSubscriber<? super T> actual;
+
+		Subscription parent;
+
+		volatile long toProduce;
```
Comment (akarnokd) [inline reactor-core/src/main/java/reactor/core/publisher/FluxLimitRequest.java]: No need for `volatile`. Or at least have a field updater and lazySet updates to it.

Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/publisher/FluxLimitRequest.java]: indeed. removed the `volatile` in latest commit 👍

**Comment by yschimke** [comment]
@simonbasle One comment about this new operator, which looks great for us BTW.  

In the case of `p.limitRate(10)`, it is unfortunate for the client to wait until credits have all been consumed before requesting more as this creates gaps, and the server may need to tear down it's connections etc, only to start them again when the 10th response item gets to the client, which then requests another 10.

Conversely, constantly refilling as each item is consumed would be wasteful and create a lot chattier protocol but would allow the server to maintain a low tide of credits.

Ideally an overloaded form of limitRate(highTide, lowTide) would allow for the request network messages to be for 10, 5, 5, 5, 5 e.g. refilling enough ahead of time to mostly avoid ever consuming all outstanding credits.

**Comment by simonbasle** [comment]
@yschimke this sounds like the internal replenishment strategy that `limitRate` (among other operators) applies, but which is hardcoded for an amount of `prefetch - (prefetch >> 2)` - and you would like to customize it?

**Comment by yschimke** [comment]
@simonbasle I'm fine with that default, it works for me.  I'm going to look into why it didn't seem to work for me.

**Comment by yschimke** [comment]
n.b. It only kicks in at 4, which isn't great.  For 3, it degrades to the 3, 3, 3, 3 behaviour.

**Comment by simonbasle** [comment]
I think this `lowTide` is indeed very complementary with `limitRequest`, which gives you a hard cap. Keep in mind that the lowest the lowTide is, the higher the probability that replenishing requests and data production can race and cause extraneous low-tide requests, and that's where `limitRequest` would be a good companion.

**Comment by simonbasle** [comment]
(fast-tracking because the review was dismissed by last "fix test + rebase" commit)

**Pull request merged**